README.md
# panel Log

To show panels in console log.
It not work on VSCode output and some anothers IDEs.

## Output

```
  Application         Version             Time Running        Load                                                                                  
  My App              1.0.0               0:00                COMPLETE                                                                              
----------------------------------------------------------------------------------------------
  Size                Length              Total               Money                                                                                 
  10                  5                   15                  $ 510.37                                                                              
----------------------------------------------------------------------------------------------
[ITEM 0] x{1 value: 3 
[ITEM 1] x{2 value: 3 
[ITEM 2] x{3 value: 3 
[ITEM 3] x{4 value: 3 
[ITEM 4] x{5 value: 3 
----------------------------------------------------------------------------------------------

```

# Install

To install using npm

```

npm i  panel-log

```

## Git repository

https://github.com/reytuty/panel-log


# Instantate 

```
var panelLog = require('panel-log') ;
    panelLog.appName = "My App";
    panelLog.appVersion = "1.0.0";
    panelLog.start() ;
```

# Example 1

```
var pannelLog = require('panel-log') ;
    pannelLog.appName = "My App";
    pannelLog.appVersion = "1.0.0";
    pannelLog.start() ;
    //you can pass real things
    setTimeout(pannelLog.setPercentComplete, 500, 0.2);
    setTimeout(pannelLog.setPercentComplete, 1000, 0.4);
    setTimeout(pannelLog.setPercentComplete, 1500, 0.6);
    setTimeout(pannelLog.setPercentComplete, 2000, 0.8);
    setTimeout(pannelLog.setPercentComplete, 2500, 1);

    //your itens

    //add into line 2 and column 2 with label TEST 1 and width 20 the value 44
    pannelLog.addItem(2, 2, "TEST 1", 20, 44);
    let count = 0;

    //add into line 2 and column 3 with label TEST 2, width 20 the value that returns by method with count external attribute
    pannelLog.addItem(2, 3, "TEST 2", 20, ()=>{
        return count++;
    })

```

# Example 2

```
var pannelLog = require('panel-log') ;
    pannelLog.appName = "My App";
    pannelLog.appVersion = "1.0.0";
    pannelLog.start() ;
    //you can pass real things
    setTimeout(pannelLog.setPercentComplete, 500, 0.2);
    setTimeout(pannelLog.setPercentComplete, 1000, 0.4);
    setTimeout(pannelLog.setPercentComplete, 1500, 0.6);
    setTimeout(pannelLog.setPercentComplete, 2000, 0.8);
    setTimeout(pannelLog.setPercentComplete, 2500, 1);
    //Your custom data
    pannelLog.onUpdate.add(()=>{
        //my list
        var myList = [{x:1, value:3},{x:2, value:3},{x:3, value:3},{x:4, value:3},{x:5, value:3}] ;
        //create a line
        var width = 20 ;
        var headers = new pannelLog.Line()
        .padding(2)
        //create a column
        .column('Size', width, [pannelLog.color.cyan])
        .column('Length', width, [pannelLog.color.cyan])
        .column('Total', width, [pannelLog.color.cyan])
        .column('Money', width, [pannelLog.color.cyan])
        .fill()
        .output();
        
        //another line withi values
        var line = new pannelLog.Line()
        .padding(2)
        .column(Math.floor(Math.random()*30), width)
        .column(myList.length, width)
        .column(15, width)
        .column("$ "+(Math.random()*20000).toFixed(2), width)
        .fill()
        .output();
        //You also can use default console.log and it will persist on screen 
        console.log("----------------------------------------------------------------------------------------------") ;
        for(var i in myList){
            var item = myList[i] ;
            console.log("[ITEM "+i+"] x{"+item.x+" value: "+item.value+" ");  
        }

    }) ;
```

CHANGELOG.md
## [1.8.2] - 2020-06-22
### FIX
- fixed memory usage realistic

## [1.8.1] - 2020-06-18
### FIX
- fixed error null point


## [1.8.0] - 2020-06-18
### ADD
- add addLineItem, no group and no headers

## [1.7.0] - 2020-06-18
### ADD
- channel to change on key press


## [1.6.1] - 2020-06-18
### FIXED
- reset, reset title too

## [1.6.0] - 2019-11-28
### ADD
- reset and removeLine


## [1.5.1] - 2019-11-28
### CHANGE
- ALIGN ITENS

## [1.4.1] - 2019-11-28
### ADD
- its possible to change layout color in method value


## [1.4.0] - 2019-11-28
### ADD
- title of group method

## [1.3.0] - 2019-11-28
### ADD
- you can create column just using easy methods to add item with label and value or method to get that value

## [1.1.0] - 2019-11-28
### ADD
- get TimeRunning and percentTotal in public method


## [0.0.5] - 2019-06-05
### FIX
- addLog Method fix

## [0.0.4] - 2019-05-20
### ADD
- Propert timeUpdateScreen to change the frequence where screen clear and update

## [0.0.3] - 2019-05-20
### FIX
- First Time Running 

## [0.0.1] - 2019-05-17
### ADD
- First version

index.js
//https://www.npmjs.com/package/clui
var BasicStopwatch = require('basic-stopwatch');
var clui = require('clui'),
    clc = require('cli-color'),
    Line = clui.Line;
var Progress = clui.Progress;
var Signal = require('signals') ;
const readline = require('readline');

function PannelLog(){
    let groupLines = new Map() ;
    let lineGroup = new Map() ;
    let maxLines = 0;
    let maxColumns = 0 ;
    let hasExtraItens = false ;
    let channelList = new Map() ;
    this.registerChannel = (channelKey, channelName)=>{
        if(channelKey.length == 1){
            channelList.set(channelKey, channelName) ;
            return ;
        }
        throw new Error("registerChannel need to char on channelKey, like a, b or c not words") ;
    }
    this.clearChannel = ()=>{
        channelList.clear() ;
    }
    this.emptySpace = {
        width:20
    };
    let bWatch = new BasicStopwatch({outputFunc:(time)=>{
        //time is HH:MM:SS.mmm => HH:MM:SS
        return  time.slice(0,8);
    }});
    bWatch.start();
    let me = this ;
    this.appName = "" ;
    this.appVersion = "";
    this.timeUpdateScreen = 500 ;
    this.forceString = (item)=>{
        return item + "" ;
    }
    this.Progress = Progress ;
    //line class
    this.Line = function(){
        let _this = this;
        let line = new Line() ;
        this.padding = (value)=>{
            line.padding(value) ;
            return _this;
        }
        this.column = (content, width = 30, color = null)=>{
            content = me.forceString( content ) ;
            line.column( content , width, color ) ;
            return _this;
        }
        this.fill = ()=>{
            line.fill() ;
            return _this ;
        }
        this.output = ()=>{
            line.output();
            return _this;
        }
    } ;
    this.newLineString = "----------------------------------------------------------------------------------------------";
    this.color = clc ;
    /**
     * Every update ecra step, after clear ecra
     */
    this.onUpdate = new Signal() ;
    let logs = [];
    var limitLog = 10 ;
    let percentTotal = 0 ;
    this.memoryUsage = 0;
    let currentChannel = null ;
    var percent = new Progress(20) ;
    //TODO: make it works
    var timeRunning = bWatch.getElapsed();
    var percentString = percent.update(0);
    this.setPercentComplete = ( p )=>{
        percentTotal = p ;
        percentString = percent.update( p ) ;
        if(percentTotal >=1){
            percentString = "COMPLETE" ;
        }
    }
    this.setChannel = (channel = null)=>{
        currentChannel = channel ;
    }
    this.getTimeRunning = ()=>{
        return timeRunning ;
    }
    this.getPercentComplete = ()=>{
        return Math.floor(( percentTotal/1)*100 );
    }
    this.setLimitLog = (limit)=>{
        limit = parseInt(limit) ;
        if(isNaN(limit)){
            return;
        }
        if(limit < 0 ){
            return ;
        }
        limitLog = limit ;
    }
    function doAdd(){
            var args = Array.from( arguments );
            logs.push(args) ;
            while(logs.length > limitLog){
                logs.shift() ;
            }
    }
    this.addLog = doAdd ;
    function updateLogEcra(){
        for(var i in logs){
            console.log.apply( null, logs[i] ) ; 
        }
    }
    function getLineMap(line){
        if( !groupLines.has(line) ){
            groupLines.set(line, new Map());
        }
        return groupLines.get(line);
    }
    
    this.addLineGroup = (line, label, color = [clc.red], channel = null)=>{
        lineGroup.set(line+channel, {label, color});
    }
    this.addLineItem = (lineIndex, columnIndex, width, methodToGetValueOrValue, channel = null)=>{
        me.addItem(lineIndex, columnIndex, null, width, methodToGetValueOrValue, channel) ;
    }
    this.resetItens = ()=>{
        groupLines.clear() ;
        lineGroup.clear();
        hasExtraItens = false ;
    }
    this.removeItem = (lineIndex, columnIndex)=>{
        groupLines.delete( lineIndex ) ;
        hasExtraItens = (groupLines.size > 0) ;
    }
    this.addItem = (lineIndex, columnIndex, label, width, methodToGetValueOrValue, channel = null)=>{
        if(maxColumns < columnIndex){
            maxColumns = columnIndex ;
        }
        if(maxLines < lineIndex){
            maxLines = lineIndex ;
        }
        var item = {label, width, methodToGetValueOrValue};
        var lineMap = getLineMap(lineIndex+channel);
        lineMap.set(columnIndex, item);
        hasExtraItens = true ;
    }
    function drawEmpty(line){
        line.padding( me.emptySpace.width );
    }
    function drawLine(lineItem, propName = "label", color){
        var hasLine = false ;
        var lineDraw = new Line();
        lineDraw.padding(2) ;
        for(var j = 0; j <= maxColumns; j++){
            if(lineItem.has(j)){
                var col = lineItem.get(j);
                if(col[propName]){
                    hasLine = true ;
                }
                var type = typeof(col[propName]);
                var c = {color:color.color};
                var val = (type=="function")? col[propName](c) : col[propName]
                lineDraw.column(val+"", col.width, c.color)
            } else {
                drawEmpty(lineDraw);
            }
        }
        if(hasLine){
            lineDraw.fill().output() ;
        }
    }
    function updateItens(channel = null){
        if(!hasExtraItens){
            return;
        }
        for(var i = 0; i <= maxLines; i++){
            if(lineGroup.has(i+channel)){
                var group = new Line()
                .padding(2)
                .column(me.newLineString, me.newLineString.length, [clc.white])
                .fill()
                .output() ;
                var labelLine = lineGroup.get(i+channel);
                var group = new Line()
                .padding(2)
                .column(labelLine.label+"", 120, labelLine.color)
                .fill()
                .output() ;
            }
            var line = getLineMap(i+channel);
            if(line.size > 0){
                //draw titles
                drawLine(line, "label", {color:[clc.cyan]})
                //draw values
                drawLine(line, "methodToGetValueOrValue", {color:[clc.white]})
            } 
        }
    }
    function updateEcra(){
        timeRunning = bWatch.getElapsed();
        var loadColor = [clc.cyan] ;
        completeString = percentString ;
        var strShortCuts = "";
        if(channelList.size > 0){
            channelList.forEach((val, key)=>{
                strShortCuts += "[ ("+key+") "+val+" ]" ;
            })
        }
        var headers = new Line()
        .padding(2)
        .column('Application', 20, [clc.cyan])
        .column('Version', 20, [clc.cyan])
        .column('Time Running', 20, [clc.cyan])
        .column('Memory', 20, [clc.cyan])
        .column('Load', 20, loadColor)
        .column('Shortcuts', (strShortCuts.length>0)?strShortCuts.length+1:16, [clc.cyan])
        .fill()
        .output() ;
        
        var usedMemory = Math.round(process.memoryUsage().rss / 1048576);
        me.memoryUsage = usedMemory*1;
        if(usedMemory > 1024){
            usedMemory =  Math.round((usedMemory/1024)*100)/100;
            usedMemory = usedMemory + " Gb";
        } else {
            usedMemory = usedMemory + " Mb";
        }
        var line = new Line()
        .padding(2)
        .column(me.appName, 20)
        .column(me.appVersion, 20)
        .column(timeRunning, 20)
        .column(usedMemory, 20)
        .column(percentString, 40)
        .column(strShortCuts.length > 0 ? strShortCuts : "no shortcuts", strShortCuts.length > 0 ? strShortCuts.length+1: 16 )
        .fill()
        .output() ;
        updateItens();
        if(currentChannel){
            updateItens(currentChannel)
        }
    }
    let intervalId = null ;
    this.start = ()=>{
        me.stop() ;
        intervalId = setInterval(()=>{
            console.clear();
            updateEcra();
            console.log( me.newLineString ) ;
            me.onUpdate.dispatch();
            updateLogEcra() ;
            console.log( me.newLineString ) ;
        }, me.timeUpdateScreen) ;
    }
    this.stop = ()=>{
        clearInterval(intervalId) ;
    }
    let pannelOn = true ;
    function setReadLineKeypress(){ 
        readline.emitKeypressEvents(process.stdin);
        try{
            process.stdin.setRawMode(true);
        } catch(e){
            //
        }
        process.stdin.on('keypress', (str, key) => {
            if (key.ctrl && key.name === 'c') {
                process.exit();
            }
            if (key.ctrl && key.name === 'o') {
                pannelOn = !pannelOn;
                (pannelOn)?me.start():me.stop() ;
            }
            if (channelList.has(key.name)) {
                //change to show vessels
                me.setChannel(key.name) ;
            }
        });
        }
        setReadLineKeypress();
}

module.exports = new PannelLog() ;


example/example1.js
var pannelLog = require('../index') ;
    pannelLog.appName = "My App";
    pannelLog.appVersion = "1.0.0";
    pannelLog.start() ;
    //you can pass real things
    setTimeout(pannelLog.setPercentComplete, 500, 0.2);
    setTimeout(pannelLog.setPercentComplete, 1000, 0.4);
    setTimeout(pannelLog.setPercentComplete, 1500, 0.6);
    setTimeout(pannelLog.setPercentComplete, 2000, 0.8);
    setTimeout(pannelLog.setPercentComplete, 2500, 1);
    //Your custom data
    function getBigData(){
        var arr = [];
        for(var i = 0; i < 10000; i++){
            arr.push({x:Math.random(), y:Math.random()});
        }
        return arr;
    }
    var infiniteArray = [];
    pannelLog.addLineGroup(2, "TESTS HERE") ;
    //make itens with method
    
    
    let count = 0;
    pannelLog.addItem(2, 2, (colorInfo)=>{
        if(count%2 == 0){
            colorInfo.color = [pannelLog.color.yellow];
        }
        return "TEST 1" ;
    }, 20, 44);
    pannelLog.addItem(2, 3, "TEST 2", 20, (colorInfo)=>{
        colorInfo.color = [pannelLog.color.blue];
        return count++;
    })


    //make lines with your own
    pannelLog.onUpdate.add(()=>{
        if(pannelLog.memoryUsage < 1200){
            //spending memory to test log
            for(var i = 0; i < 100; i++){
                infiniteArray.push({x:Math.random(), y:Math.random(), i, data:getBigData()});
            }
        }
        //my list
        var myList = [{x:1, value:3},{x:2, value:3},{x:3, value:3},{x:4, value:3},{x:5, value:3}] ;
        //create a line
        var width = 20 ;
        var headers = new pannelLog.Line()
        .padding(2)
        //create a column
        .column('Size', width, [pannelLog.color.cyan])
        .column('Length', width, [pannelLog.color.cyan])
        .column('Total', width, [pannelLog.color.cyan])
        .column('Money', width, [pannelLog.color.cyan])
        .fill()
        .output();
        
        //another line withi values
        var line = new pannelLog.Line()
        .padding(2)
        .column(Math.floor(Math.random()*30), width)
        .column(myList.length, width)
        .column(15, width)
        .column("$ "+(Math.random()*20000).toFixed(2), width)
        .fill()
        .output();
        //You also can use default console.log and it will persist on screen 
        console.log("----------------------------------------------------------------------------------------------") ;
        for(var i in myList){
            var item = myList[i] ;
            console.log("[ITEM "+i+"] x{"+item.x+" value: "+item.value+" ");  
        }

    }) ;